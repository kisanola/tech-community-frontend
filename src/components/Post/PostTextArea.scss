@import '../../styles/colors';

#text-area-post {
  width: 100%;
  display: flex;
  flex-direction: column;
  position: relative;

  background-color: $secondary;

  ::-webkit-input-placeholder {
    color: $surface;
  }
  ::-moz-placeholder {
    color: $surface;
  }
  :-ms-input-placeholder {
    color: $surface;
  }
  :-moz-placeholder {
    color: $surface;
  }

  .text-area-post {
    width: 100%;
    min-width: 400px;
    max-width: 100%;
    padding: 20px;
    font-size: 1rem;
    letter-spacing: 1px;
    color: $surface;
    border: none;
    outline: none;
    background-color: transparent;
    overflow: auto;
    resize: none;
  }

  .actions {
    display: flex;
    justify-content: space-between;
    align-items: center;
    width: 100%;
    padding: 0rem 1rem 1.2rem 1rem;
    position: 'relative';
    i {
      color: $surface;
      margin: 0 0.5em;
      font-size: 1.5em;
      cursor: pointer;
    }
    span {
      color: $surface;
      margin: 0 0.5em;
      font-size: 1.5em;
      cursor: pointer;
    }
  }

  .post-action {
    display: flex;
    flex-direction: row;
    align-items: center;

    .number-character {
      font-size: 0.8em;
      margin-right: 1rem;
    }
  }

  #emoji {
    width: 0px;
    border: none;
    background-color: transparent;
    outline: none;
  }

  .img-preview {
    height: 25px;
    width: 30px;
    margin: 0 0.5em;
    vertical-align: sub;
  }
}
