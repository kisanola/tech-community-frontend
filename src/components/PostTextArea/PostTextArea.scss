@import '../../styles/colors';

#text-area-post {
  margin-left: 27%;
  margin-right: 27%;
  width: 100%;
  display: flex;
  flex-direction: column;
  position: relative;

  background-color: $secondary;

  ::-webkit-input-placeholder {
    color: $surface;
  }
  ::-moz-placeholder {
    color: $surface;
  }
  :-ms-input-placeholder {
    color: $surface;
  }
  :-moz-placeholder {
    color: $surface;
  }

  .text-area-post {
    width: 100%;
    min-width: 400px;
    max-width: 100%;
    padding: 20px;
    font-size: 1rem;
    letter-spacing: 1px;
    color: $surface;
    border: none;
    outline: none;
    background-color: transparent;
    overflow: auto;
    resize: none;
  }

  .actions {
    display: flex;
    justify-content: space-between;
    width: 100%;
    padding: 0rem 1rem 1.2rem 1rem;
    position: 'relative';

    i {
      color: $surface;
      margin: 0 0.5em;
      font-size: 1.5em;
      cursor: pointer;
    }

    span {
      color: $surface;
      margin: 0 0.5em;
      font-size: 1.5em;
      cursor: pointer;
    }
  }
  
  .img-preview {
    height: 25px;
    width: 30px;
    margin: 0 0.5em;
    vertical-align: sub;
  }
}
