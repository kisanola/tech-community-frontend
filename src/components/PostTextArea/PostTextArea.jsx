import React, { Component } from 'react';
import { Picker } from 'emoji-mart';
import PropTypes from 'prop-types';
import Button from '../Button/Button';
import Modal from '../Modal/Modal';
import 'emoji-mart/css/emoji-mart.css';
import './PostTextArea.scss';

const buttonStyle = {
  paddingTop: '3px',
  paddingBottom: '3px',
  paddingLeft: '10px',
  paddingRight: '10px',
  fontSize: '0.6em',
};

const categoryButtonStyle = {
  paddingTop: '3px',
  paddingBottom: '3px',
  paddingLeft: '10px',
  paddingRight: '10px',
  fontSize: '1rem',
  width: '100%',
};

export class PostTextArea extends Component {
  state = {
    value: '',
    rows: 3,
    show: false,
    showEmojiPicker: false,
    disabled: true,
    showConfirmCategory: false,
    category: 'general',
  };

  imageInput = React.createRef();

  emojiToggle = React.createRef();

  componentDidMount() {
    window.addEventListener('click', (e) => {
      const {
        target: { id, parentNode },
      } = e;
      let isEmoji = false;
      let node = parentNode;
      for (let i = 0; i < 6; i += 1) {
        if (!node) break;
        if (!node.classList) break;
        isEmoji = node && node.classList.contains('emoji-mart');
        node = node.parentNode;
        if (isEmoji) break;
      }
      if (id !== 'emoji-btn' && !isEmoji) {
        this.setState({
          showEmojiPicker: false,
        });
      }
    });
  }

  componentDidUpdate = (prevProps) => {
    const { tick } = this.props;
    if (prevProps.tick !== tick) {
      this.setState({
        value: '',
        rows: 3,
        showEmojiPicker: false,
        imageUrl: '',
        image: {},
      });
    }
  };

  trimedTextLength = (text) => String(text).replace(/\s\s+/g, ' ').length;

  _onChange = (e) => {
    const textareaLineHeight = 24;
    const text = e.target.value;
    const textLength = this.trimedTextLength(text);

    const {
      minRows, maxRows, minChar, maxChar, onChange,
    } = this.props;
    if (maxChar && maxChar <= e.target.value.length - 1) return;

    if (textLength >= minChar) this.setState({ disabled: false });
    if (textLength < minChar) this.setState({ disabled: true });

    const previousRows = e.target.rows;
    e.target.rows = minRows; // reset number of rows in textarea
    const currentRows = ~~(e.target.scrollHeight / textareaLineHeight);

    if (currentRows === previousRows) {
      e.target.rows = currentRows;
    }

    if (currentRows >= maxRows) {
      e.target.rows = maxRows;
      e.target.scrollTop = e.target.scrollHeight;
    }

    this.setState({
      value: e.target.value,
      rows: currentRows < maxRows ? currentRows : maxRows,
    });

    if (onChange) onChange(e.target.value);
  };

  onImageChange = (e) => {
    this.setState({
      imageUrl: URL.createObjectURL(e.target.files[0]),
      image: e.target.files[0],
    });
  };

  onCategoryChange = ({ target: { id } }) => {
    this.setState({
      category: id,
    });
  }

  restoreImage = () => {
    this.setState({
      imageUrl: '',
      image: {},
    });
  };

  addEmoji = (emoji) => {
    const { minChar } = this.props;
    const { value: textValue } = this.state;
    const value = textValue + emoji.native;

    this.setState({ value });
    if (value.length >= minChar) this.setState({ disabled: false });
  };

  toogleOnClickEmojiPicker = () => {
    const { showEmojiPicker } = this.state;
    this.setState({ showEmojiPicker: !showEmojiPicker });
  };

  showModal = () => {
    this.setState({ show: true });
  };

  hideModal = () => {
    this.setState({ show: false });
  };

  submitPost = () => {
    const {
      onToggleSocialModal, isAuth, post, slug, comment,
    } = this.props;
    if (!isAuth) return onToggleSocialModal(true);

    const {
      value, image = '', showConfirmCategory, category,
    } = this.state;

    if (value && (showConfirmCategory || comment)) {
      post({
        value, image, slug, category,
      });
      this.setState({
        showConfirmCategory: false,
      });
      return;
    }
    this.setState({
      showConfirmCategory: true,
    });
  };

  closeConfirmCategory =() => {
    this.setState({
      showConfirmCategory: false,
    });
  };

  renderConfirmCategory = () => {
    const { showConfirmCategory, category } = this.state;
    const {
      loading,
      tick,
    } = this.props;
    return (
      <Modal show={showConfirmCategory} handleClose={this.closeConfirmCategory}>
        <div className="post-category">
          <h3>Choose post category</h3>
          <p>
            <input
              type="radio"
              id="general"
              name="category"
              onChange={this.onCategoryChange}
              checked={category === 'general'}
            />
            <label htmlFor="general">
              <span>
                <i className="fa fa-globe" />
              </span>
              General
            </label>
          </p>
          <p>
            <input
              type="radio"
              id="job"
              name="category"
              onChange={this.onCategoryChange}
              checked={category === 'job'}
            />
            <label htmlFor="job">
              <span>
                <i className="fa fa-suitcase" />
              </span>
              Jobs
            </label>
          </p>
          <p>
            <input
              type="radio"
              id="event"
              name="category"
              onChange={this.onCategoryChange}
              checked={category === 'event'}
            />
            <label htmlFor="event">
              <span>
                <i className="fa fa-calendar" />
              </span>
              Events
            </label>
          </p>
          <p>
            <input
              type="radio"
              id="annoucement"
              name="category"
              onChange={this.onCategoryChange}
              checked={category === 'annoucement'}
            />
            <label htmlFor="annoucement">
              <span>
                <i className="fa fa-bullhorn" />
              </span>
              Announcements
            </label>
          </p>

          <p>
            <input
              type="radio"
              id="question"
              name="category"
              onChange={this.onCategoryChange}
              checked={category === 'question'}
            />
            <label htmlFor="question">
              <span>
                <i className="fa fa-question" />
              </span>
              Questions
            </label>
          </p>

          <Button
            style={categoryButtonStyle}
            loading={loading}
            text="POST"
            tick={tick}
            onClick={() => this.submitPost()}
          />
        </div>
      </Modal>
    );
  }

  render() {
    const {
      showEmojiPicker,
      value,
      rows,
      imageUrl,
      disabled,
      show,
    } = this.state;

    const textLength = this.trimedTextLength(value);

    const {
      loading,
      tick,
      maxChar,
      allowImagePicker,
      placeholder,
    } = this.props;
    return (
      <>
        {this.renderConfirmCategory()}
        <div id="text-area-post">
          <textarea
            className="text-area-post"
            value={value}
            rows={rows}
            onChange={this._onChange}
            placeholder={placeholder}
          />

          <input
            type="file"
            name="image"
            className="file-upload"
            onChange={this.onImageChange}
            accept="image/gif, image/jpeg, image/png, image/jpg"
            ref={this.imageInput}
            style={{ display: 'none' }}
          />
          <Modal show={show} handleClose={this.hideModal}>
            <img
              role="presentation"
              className="img-preview"
              style={{ width: '70%' }}
              alt="preview"
              src={imageUrl}
              onClick={this.showModal}
            />
          </Modal>
          {imageUrl ? (
            <div className="image">
              <div
                role="presentation"
                className="remove-image"
                onClick={this.restoreImage}
              >
                <span>&times;</span>
              </div>
              <img
                role="presentation"
                className="img-preview"
                alt="preview"
                src={imageUrl}
                onClick={this.showModal}
              />
            </div>
          ) : null}
          <div className="actions">
            <div>
              <i
                role="presentation"
                id="emoji-btn"
                className="fas fa-smile"
                onClick={this.toogleOnClickEmojiPicker}
              />
              {allowImagePicker && (
              <i
                role="presentation"
                className="fas fa-image"
                onClick={() => this.imageInput.current.click()}
              />
              )}

              {showEmojiPicker && (
              <Picker
                title="Tech Community"
                emoji="point_up"
                style={{ position: 'absolute', left: 0, top: '100%' }}
                onSelect={this.addEmoji}
              />
              )}
            </div>
            <div
              title={disabled ? 'The minimum number of character is 50' : ''}
              className="post-action"
            >
              <span
                className="number-character"
                style={{ color: disabled ? '#fff' : '#13c39a' }}
              >
                {textLength}
                /
                {maxChar}
              </span>
              <Button
                style={buttonStyle}
                loading={loading}
                text="POST"
                disabled={disabled}
                tick={tick}
                onClick={() => this.submitPost()}
              />
            </div>
          </div>
        </div>
      </>
    );
  }
}

PostTextArea.propTypes = {
  minRows: PropTypes.number,
  maxRows: PropTypes.number,
  minChar: PropTypes.number,
  maxChar: PropTypes.number,
  slug: PropTypes.string,
  allowImagePicker: PropTypes.bool,
  placeholder: PropTypes.string,
  loading: PropTypes.bool,
  tick: PropTypes.bool,
  post: PropTypes.func.isRequired,
  onChange: PropTypes.func,
  comment: PropTypes.bool,
  isAuth: PropTypes.bool.isRequired,
  onToggleSocialModal: PropTypes.func.isRequired,
};

PostTextArea.defaultProps = {
  minRows: 3,
  maxRows: 15,
  minChar: 50,
  maxChar: 500,
  slug: null,
  allowImagePicker: true,
  placeholder: 'Write something ...',
  loading: false,
  tick: false,
  onChange: () => '',
  comment: false,
};

export default PostTextArea;
