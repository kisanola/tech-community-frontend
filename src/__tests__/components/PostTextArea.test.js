import React from 'react';
import { shallow, mount } from 'enzyme';
import PostTextArea from '../../components/PostTextArea/PostTextArea';

let wrapper;

const props = {
  minRows: 3,
  maxRows: 15,
  loading: false,
  error: false,
  tick: false,
  post: jest.fn(),
};

describe('PostTextArea.jsx', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('should render PostTextArea.jx', () => {
    wrapper = shallow(<PostTextArea {...props} />);
    const event = {
      target: {
        value: 'this is the PostTextArea',
      },
    };
    wrapper.find('textarea').simulate('change', event);
    expect(wrapper.instance().state.value).toEqual(event.target.value);
  });

  test('should PostTextArea', () => {
    const newProps = { ...props, allowImagePicker: false };
    wrapper = shallow(<PostTextArea {...newProps} />);
    const event = { target: { value: 'this is the PostTextArea' } };
    wrapper.find('textarea').simulate('change', event);
    wrapper.find('Button').simulate('click');

    expect(newProps.post).toHaveBeenCalled();
  });

  test('should select the emage', () => {
    wrapper = shallow(<PostTextArea {...props} />);
    global.URL.createObjectURL = jest.fn(() => 'this is the imageUrl');
    const event = { target: { files: ['this is the image'] } };
    wrapper.find('input[name="image"]').simulate('change', event);
    expect(global.URL.createObjectURL).toHaveBeenCalled();
    expect(wrapper.instance().state.image).toEqual(event.target.files[0]);
  });

  test('should clear the image', () => {
    wrapper = shallow(<PostTextArea {...props} />);
    global.URL.createObjectURL = jest.fn(() => 'this is the imageUrl');
    const event = { target: { files: ['this is the image'] } };
    wrapper.find('input[name="image"]').simulate('change', event);
    wrapper.find('.remove-image').simulate('click');
    expect(wrapper.instance().state.imageUrl).toEqual('');
  });

  test('should add emoji', () => {
    wrapper = mount(
      <PostTextArea {...props} />,
    );
    const emoji = { native: '&123' };
    wrapper.instance().addEmoji(emoji);
    expect(wrapper.instance().state.value).toEqual('&123');
  });

  test('should show emoji picker', () => {
    wrapper = mount(
      <PostTextArea {...props} />,
    );
    wrapper.instance().toogleOnClickEmojiPicker();
    expect(wrapper.instance().state.showEmojiPicker).toEqual(true);
  });

  test('should hide emoji picker', () => {
    wrapper = mount(
      <PostTextArea {...props} />,
    );
    const event = {
      target: { id: 'id', parentNode: { classList: 'this is the classList' } },
    };
    wrapper.find('button').simulate('click', event);
    expect(wrapper.instance().state.showEmojiPicker).toEqual(false);
  });

  test('should show the modal', () => {
    wrapper = mount(
      <PostTextArea {...props} />,
    );
    wrapper.instance().showModal();
    expect(wrapper.instance().state.show).toEqual(true);
  });

  test('should hide the modal', () => {
    wrapper = mount(
      <PostTextArea {...props} />,
    );
    wrapper.instance().hideModal();
    expect(wrapper.instance().state.show).toEqual(false);
  });
});
