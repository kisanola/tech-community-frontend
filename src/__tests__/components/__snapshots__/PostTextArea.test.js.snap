// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostTextArea.jsx should render PostTextArea.jx 1`] = `ShallowWrapper {}`;

exports[`PostTextArea.jsx should render PostTextArea.jx 2`] = `ShallowWrapper {}`;
